os/fs/smartfs/smartfs_utils.c

```diff
@@ -1716,139 +1963,263 @@ int smartfs_countdirentries(struct smartfs_mountpt_s *fs, struct smartfs_entry_s
 }
 
 /****************************************************************************
- * Name: smartfs_truncatefile
+ * Name: smartfs_shrinkfile
  *
- * Description: Truncates an existing file on the device so that it occupies
- *              zero bytes and can be completely re-written.
+ * Description:
+ *   Shrink the size of an existing file to the specified length
  *
  ****************************************************************************/
 
-int smartfs_truncatefile(struct smartfs_mountpt_s *fs, struct smartfs_entry_s *entry, FAR struct smartfs_ofile_s *sf)
+/* NOTE:- This function deviates slightly from the POSIX standard as implemented in Linux.
+ *        The file position pointer in Linux is not reset to the new data length. If a write
+ *        operation is requested after shrink, new data is written starting at the old file
+ *        position pointer. The extra space in between gets filled up with '0's.
+ *        Smartfs shrinkfile function will reset the file pointer to the position of 'length'
+ *        as requested in the shrink operation. New data is written starting at this position.
+ *        Any conflicts/issues arising due to this deviation will be taken care of.
+ */
+int smartfs_shrinkfile(FAR struct smartfs_mountpt_s *fs, FAR struct smartfs_ofile_s *sf, off_t length)
 {
 	int ret;
-	uint16_t nextsector;
-	uint16_t sector;
-	struct smartfs_chain_header_s *header;
 	struct smart_read_write_s readwrite;
+	struct smartfs_chain_header_s *chainheader;
+	uint16_t sector;
+	uint16_t nextsector;
+	uint16_t data_len;
+	uint16_t offset;
+	uint16_t last_data;
+	uint16_t savefilepos;
+
+	/* 'data_len' represents length of DATA in a sector, the chainheader is not counted in file data */
+	data_len = fs->fs_llformat.availbytes - sizeof(struct smartfs_chain_header_s);
+	/* Number of bytes to be retained in the last sector after shrinking file */
+	last_data = (length % data_len);
```

Comment:
I think remained_data is more good, how do you think Mr. native(@amogh-samsung) ?

Reply:
:P
Yes, this variable name is a tough choice, difficult to do in 2 words.....
`Remained_data `or `Remaining_data `does not clarify that this is the data length in the last sector of the file.
How about `last_sector_data`?

Reply:
Alignment with smartfs_write is a consideration but `byteswritten` will be totally different from what the variable means.
Would you prefer `last_sect_data `if not `last_sector_data`?